File: CSharp/SpaceVIL/SpaceVIL/Items/ComboBoxDropDown.cs
using System;
using System.Drawing;
using System.Collections.Generic;
using SpaceVIL.Core;
using SpaceVIL.Common;
using SpaceVIL.Decorations;

namespace SpaceVIL
{
    public class ComboBoxDropDown : Prototype, IFloating
    {
        public EventCommonMethod SelectionChanged;
        public Prototype ReturnFocus = null;
        public ListBox ItemList = new ListBox();
        private String _text_selection = String.Empty;
        public String GetText()
        {
            return _text_selection;
        }
        public int GetCurrentIndex()
        {
            return ItemList.GetSelection();
        }
        public void SetCurrentIndex(int index)
        {
            if (!_init)
                InitElements();

            ItemList.SetSelection(index);
            MenuItem selection = ItemList.GetSelectionItem() as MenuItem;
            if (selection != null)
                _text_selection = selection.GetText();
        }
        private Queue<IBaseItem> _queue = new Queue<IBaseItem>();

        private static int count = 0;
        public MouseButton ActiveButton = MouseButton.ButtonLeft;

        private bool _init = false;
        private bool _ouside = true;

        /// <summary>
        /// Close the ComboBoxDropDown it mouse click is outside (true or false)
        /// </summary>
        public bool IsOutsideClickClosable()
        {
            return _ouside;
        }
        public void SetOutsideClickClosable(bool value)
        {
            _ouside = value;
        }
        // private bool _lock_ouside = true;
        // public bool IsLockOutside()
        // {
        //     return _lock_ouside;
        // }
        // public void SetLockOutside(bool value)
        // {
        //     _lock_ouside = value;
        // }

        /// <summary>
        /// Constructs a ComboBoxDropDown
        /// </summary>
        /// <param name="handler"> parent window for the ComboBoxDropDown </param>
        public ComboBoxDropDown(WindowLayout handler)
        {
            SetPassEvents(false);
            SetVisible(false);
            SetHandler(handler);
            SetItemName("ComboBoxDropDown_" + count);
            count++;
            ItemsLayoutBox.AddItem(GetHandler(), this, LayoutType.Floating);
            SetStyle(DefaultsService.GetDefaultStyle(typeof(SpaceVIL.ComboBoxDropDown)));
            SetShadow(5, 3, 3, Color.FromArgb(180, 0, 0, 0));
        }

        /// <summary>
        /// Initialization and adding of all elements in the ComboBoxDropDown
        /// </summary>
        public override void InitElements()
        {
            SetConfines();
            ItemList.SetVScrollBarVisible(ScrollBarVisibility.AsNeeded);
            ItemList.SetHScrollBarVisible(ScrollBarVisibility.AsNeeded);
            ItemList.GetArea().SelectionChanged += OnSelectionChanged;

            base.AddItem(ItemList);

            foreach (var item in _queue)
                ItemList.AddItem(item);
            _queue = null;
            _init = true;

            ItemList.GetArea().EventKeyPress += (sender, args) =>
            {
                if (args.Key == KeyCode.Escape)
                    Hide();
            };
        }

        void OnSelectionChanged()
        {
            MenuItem selection = ItemList.GetSelectionItem() as MenuItem;
            if (selection != null)
                _text_selection = selection.GetText();
            Hide();
            SelectionChanged?.Invoke();
        }

        /// <summary>
        /// Returns count of the ComboBoxDropDown lines
        /// </summary>
        public int GetListCount()
        {
            return ItemList.GetListContent().Count;
        }

        /// <summary>
        /// Returns ComboBoxDropDown items list
        /// </summary>
        public List<IBaseItem> GetListContent()
        {
            return ItemList.GetListContent();
        }

        /// <summary>
        /// Add item to the ComboBoxDropDown
        /// </summary>
        public override void AddItem(IBaseItem item)
        {
            _queue.Enqueue(item);
        }

        /// <summary>
        /// Remove item from the ComboBoxDropDown
        /// </summary>
        public override void RemoveItem(IBaseItem item)
        {
            ItemList.RemoveItem(item);
        }

        void UpdateSize()
        {
            int height = 0;
            int width = GetWidth();
            List<IBaseItem> list = ItemList.GetListContent();
            foreach (var item in list)
            {
                height += (item.GetHeight() + ItemList.GetArea().GetSpacing().Vertical);

                int tmp = GetPadding().Left + GetPadding().Right + item.GetMargin().Left + item.GetMargin().Right;

                MenuItem m = item as MenuItem;
                if (item != null)
                {
                    tmp += m.GetTextWidth() + m.GetMargin().Left + m.GetMargin().Right + m.GetPadding().Left + m.GetPadding().Right;
                }
                else
                    tmp = tmp + item.GetWidth() + item.GetMargin().Left + item.GetMargin().Right;

                if (width < tmp)
                    width = tmp;
            }
            // SetSize(width, height);
            SetWidth(width);
            SetHeight(height);
        }

        /// <summary>
        /// Show the ComboBoxDropDown
        /// </summary>
        /// <param name="sender"> the item from which the show request is sent </param>
        /// <param name="args"> mouse click arguments (cursor position, mouse button,
        /// mouse button press/release, etc.) </param>
        public void Show(IItem sender, MouseArgs args)
        {
            if (args.Button == ActiveButton)
            {
                if (!_init)
                    InitElements();

                SetVisible(true);
                SetConfines();
                ItemList.GetArea().SetFocus();
            }
        }

        /// <summary>
        /// Hide the ComboBoxDropDown without destroying
        /// </summary>
        public void Hide()
        {
            SetX(-GetWidth());
            SetVisible(false);
            // ItemList.Unselect();
            ReturnFocus?.SetFocus();
        }

        /// <summary>
        /// Set confines according to position and size of the ComboBoxDropDown
        /// </summary>
        public override void SetConfines()
        {
            base.SetConfines(
                GetX(),
                GetX() + GetWidth(),
                GetY(),
                GetY() + GetHeight()
            );
        }

        //style
        /// <summary>
        /// Set style of the ComboBoxDropDown
        /// </summary>
        public override void SetStyle(Style style)
        {
            if (style == null)
                return;
            SetPadding(style.Padding);
            SetSizePolicy(style.WidthPolicy, style.HeightPolicy);
            SetBackground(style.Background);

            Style inner_style = style.GetInnerStyle("itemlist");
            if (inner_style != null)
            {
                ItemList.SetStyle(inner_style);
            }
            inner_style = style.GetInnerStyle("listarea");
            if (inner_style != null)
            {
                ItemList.GetArea().SetStyle(inner_style);
            }
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/ComboBoxDropDown.java
package com.spvessel.spacevil;

import java.awt.Color;
import java.util.LinkedList;
import java.util.List;

import com.spvessel.spacevil.Common.DefaultsService;
import com.spvessel.spacevil.Core.EventCommonMethod;
import com.spvessel.spacevil.Core.InterfaceBaseItem;
import com.spvessel.spacevil.Core.InterfaceFloating;
import com.spvessel.spacevil.Core.InterfaceItem;
import com.spvessel.spacevil.Core.MouseArgs;
import com.spvessel.spacevil.Decorations.Style;
import com.spvessel.spacevil.Flags.KeyCode;
import com.spvessel.spacevil.Flags.LayoutType;
import com.spvessel.spacevil.Flags.MouseButton;
import com.spvessel.spacevil.Flags.ScrollBarVisibility;

public class ComboBoxDropDown extends Prototype implements InterfaceFloating {
    public EventCommonMethod selectionChanged = new EventCommonMethod();
    public Prototype returnFocus = null;
    public ListBox itemList = new ListBox();
    private String _text_selection = "";

    public String getText() {
        return _text_selection;
    }

    public int getCurrentIndex() {
        return itemList.getSelection();
    }

    public void setCurrentIndex(int index) {
        if (!_init)
            initElements();

        itemList.setSelection(index);
        if (itemList.getSelectionItem() instanceof MenuItem) {
            MenuItem selection = (MenuItem) itemList.getSelectionItem();
            _text_selection = selection.getText();
        }
    }

    private List<InterfaceBaseItem> _queue = new LinkedList<>();

    private static int count = 0;
    public MouseButton activeButton = MouseButton.BUTTON_LEFT;

    private boolean _init = false;
    private boolean _ouside = true;

    /**
     * Close the ContextMenu it mouse click is outside (true or false)
     */
    public boolean isOutsideClickClosable() {
        return _ouside;
    }

    public void setOutsideClickClosable(boolean value) {
        _ouside = value;
    }

    /**
     * Constructs a ContextMenu
     * 
     * @param handler parent window for the ContextMenu
     */
    public ComboBoxDropDown(WindowLayout handler) {
        setPassEvents(false);
        setVisible(false);
        setHandler(handler);
        setItemName("ComboBoxDropDown_" + count);
        count++;
        ItemsLayoutBox.addItem(getHandler(), this, LayoutType.FLOATING);
        setStyle(DefaultsService.getDefaultStyle(ComboBoxDropDown.class));
        setShadow(5, 3, 3, new Color(0, 0, 0, 180));
    }

    /**
     * Initialization and adding of all elements in the ContextMenu
     */
    @Override
    public void initElements() {
        setConfines();
        itemList.setVScrollBarVisible(ScrollBarVisibility.AS_NEEDED);
        itemList.setHScrollBarVisible(ScrollBarVisibility.AS_NEEDED);
        itemList.getArea().selectionChanged.add(this::onSelectionChanged);

        super.addItem(itemList);

        for (InterfaceBaseItem item : _queue) {
            itemList.addItem(item);
        }
        _queue = null;
        _init = true;

        itemList.getArea().eventKeyPress.add((sender, args) -> {
            if (args.key == KeyCode.ESCAPE) {
                hide();
            }
        });
        updateSize();
    }

    private void onSelectionChanged() {
        if (itemList.getSelectionItem() instanceof MenuItem) {
            MenuItem selection = (MenuItem) itemList.getSelectionItem();
            _text_selection = selection.getText();
        }
        hide();
        selectionChanged.execute();
    }

    /**
     * Returns count of the ContextMenu lines
     */
    public int getListCount() {
        return itemList.getListContent().size();
    }

    /**
     * Returns ContextMenu items list
     */
    public List<InterfaceBaseItem> getListContent() {
        return itemList.getListContent();
    }

    /**
     * Add item to the ContextMenu
     */
    @Override
    public void addItem(InterfaceBaseItem item) {
        _queue.add(item);
    }

    /**
     * Remove item from the ContextMenu
     */
    @Override
    public void removeItem(InterfaceBaseItem item) {
        itemList.removeItem(item);
    }

    private void updateSize() {
        int height = 0;
        int width = getWidth();
        List<InterfaceBaseItem> list = itemList.getListContent();
        for (InterfaceBaseItem item : list) {
            InterfaceBaseItem wrapper = itemList.getWrapper(item);
            height += (wrapper.getHeight() + itemList.getArea().getSpacing().vertical);

            int tmp = getPadding().left + getPadding().right + item.getMargin().left + item.getMargin().right;

            if (item instanceof MenuItem) {
                MenuItem m = (MenuItem) item;
                tmp += m.getTextWidth() + m.getMargin().left + m.getMargin().right + m.getPadding().left
                        + m.getPadding().right;
            } else
                tmp = tmp + item.getWidth() + item.getMargin().left + item.getMargin().right;

            if (width < tmp)
                width = tmp;
        }
        // SetSize(width, height);
        setWidth(width);
        setHeight(height);
    }

    /**
     * Show the ContextMenu
     * 
     * @param sender the item from which the show request is sent
     * @param args   mouse click arguments (cursor position, mouse button, mouse
     *               button press/release, etc.)
     */
    public void show(InterfaceItem sender, MouseArgs args) {
        if (args.button.getValue() == activeButton.getValue()) {
            if (!_init)
                initElements();
            setVisible(true);
            setConfines();
            itemList.getArea().setFocus();
        }
    }

    /**
     * Hide the ContextMenu without destroying
     */
    public void hide() {
        setX(-getWidth());
        setVisible(false);
        // itemList.Unselect();
        if (returnFocus != null)
            returnFocus.setFocus();
    }

    /**
     * Set confines according to position and size of the ContextMenu
     */
    @Override
    public void setConfines() {
        super.setConfines(getX(), getX() + getWidth(), getY(), getY() + getHeight());
    }

    /**
     * Set style of the ContextMenu
     */
    @Override
    public void setStyle(Style style) {
        if (style == null)
            return;
        setPadding(style.padding);
        setSizePolicy(style.widthPolicy, style.heightPolicy);
        setBackground(style.background);

        Style inner_style = style.getInnerStyle("itemlist");
        if (inner_style != null) {
            itemList.setStyle(inner_style);
        }
        inner_style = style.getInnerStyle("listarea");
        if (inner_style != null) {
            itemList.getArea().setStyle(inner_style);
        }
    }
}